geo/src/algorithm/line_locate_point.rs

```diff
@@ -85,12 +90,28 @@ where
             ));
             total_length += length;
         }
-        let l = queue
-            .iter()
-            .min_by(|x, y| (x.2).partial_cmp(&y.2).unwrap())
-            .unwrap();
-
-        (l.0 + l.1 * (l.3).line_locate_point(p)) / total_length
+        if total_length == T::zero() {
+            // linestring has zero legnth, return zero
+            return T::zero();
```

Comment:
Am I correct that giving the start point of the line gives me the same answer as giving some random point which isn't on the line?

Reply:
I hope I'm interpreting your question correctly - for a zero length line, yes you'll get zero as the returned fraction for any point in the plane (that's the idea of the test on line 189).

For a non-zero length line, the fold statement below finds the first line segment that is closest to the point, and returns the fractional length of the linestring up to the closest point on that line segment - test on line 182 demonstrates this - the point is equidistant to the first and last line segment, but defers to the first.

Does that make sense/answer your question?

Reply:
Sorry, my question wasn't very clear.

```
let line_string = line_string!((0.0, 0.0), (10.0, 0.));

// this makes sense to me, we are on the line and 50% along it.
assert_eq!(0.5, line_string.line_locate_point(Point::new(0.0, 5.0)))

// this makes sense to me, we are on the line but 0% along it.
assert_eq!(0.0, line_string.line_locate_point(Point::new(0.0, 0.0)))

// I'm a little surprised that this case is indistinguishable from the previous, since we're not even on the line.
assert_eq!(0.0, line_string.line_locate_point(Point::new(555.0, 555.0))
```

Provided I'm interpreting this correctly, I'm wondering if it would be better to, similar to my other comment, treat the case of "off the line" as some kind of explicit failure, either by returning a Result type or an Option type.

An example of where I could imagine this being surprising:
```
let line_string = line_string!((0.0, 0.0), (10.0, 0.));

assert_eq!(0.0, line_string.line_locate_point(Point::new(0.0, 0.0)))
assert_eq!(0.9, line_string.line_locate_point(Point::new(0.0, 9.0)))
assert_eq!(0.99, line_string.line_locate_point(Point::new(0.0, 9.9)))
assert_eq!(0.999, line_string.line_locate_point(Point::new(0.0, 9.99)))
assert_eq!(1.0, line_string.line_locate_point(Point::new(0.0, 10.0)))

# is this surprising?
assert_eq!(0.0, line_string.line_locate_point(Point::new(0.0, 10.000001)))

# or given this
assert_eq!(0.5, line_string.line_locate_point(Point::new(0.0, 5.0)))
# is this suprising?
assert_eq!(0.0, line_string.line_locate_point(Point::new(0.000001, 5.0)))
```

Granted it's not *that* surprising in a unit test like my example, but if you're building the input point from some sequence of operations where floating point error can accumulate, it seems possible that we'd slip off the line just a bit.

The previous `unwrap` is brutal, but it helped keep the developer from shooting themselves in the foot unknowingly. Maybe explicitly signaling "not on the line" (with an Option or a Result) would be a good compromise.

What do you think @mattswoon?

Reply:
Right, that makes sense. My intention with these two traits is to help you locate how far one would travel along a line in order to get the point on the line that is _nearest to the given point_, so should return a fraction for any finite point in the plane.  This is the behaviour from postgis, which is why I kept the same names too (https://postgis.net/docs/ST_LineLocatePoint.html).

In your first example, the last assert would actually be `assert_eq!(1.0, line_string.line_locate_point(Point::new(555.0, 555.0))` as the nearest point on `line_string` to `(555, 555)` is the end point.

So my apologies for the confusion, the fraction doesn't "default" to 0.0 for points off the line - below is my crude drawing for what you'd expect for a single line segment.

![a single line segment](https://user-images.githubusercontent.com/10151166/95922951-234d7600-0e00-11eb-8693-80ec8239baf5.jpg).

For the linestring case, things are slightly more complicated because the point off the line can be equidistant to multiple line segments, and in this case, the fraction return is that for the first, closest line segment. I've got another picture:

![linestring](https://user-images.githubusercontent.com/10151166/95923374-1aa96f80-0e01-11eb-9a2c-8ceeab3cc70b.jpg)

The green and pink dots are equidistant to the first/second and first/third line segments respectively, so the fraction returned is that of the projection to the first line segment. The purple dot is "very close" to the green dot, but closer to the third line segment than the first, so the fraction return is quite a bit larger. But I think this ok.

I'm actually just about to push changes where these will return options, and result in `None` when either the line has nan/infinite coordinates, the point has nan/infinite coordinates for LineLocatePoint, or the fraction is nan for LineInterpolatePoint (interpolating infinitely along a line just gives the end point, and negative-infinitely the start point).

Reply:
Thank you for taking the time to explain this.